### PR TITLE
pd-client:  remove `call_option` to avoid  deadlock(RWR).

### LINKED ISSUE #13191: Inject a minority of tikv network isolation，after network isolation was recovered, one tikv status still show down and can not transfer leader back
### What version of TiKV are you using?
./tikv-server -V
 TiKV 
Release Version:   6.2.0-alpha
Edition:           Community
Git Commit Hash:   940e1395869e2d92aa91eb2d59380ce894125b70
Git Commit Branch: heads/refs/tags/v6.2.0-alpha
UTC Build Time:    2022-07-29 11:04:30
Rust Version:      rustc 1.62.0-nightly (7c4b47696 2022-04-30)
Enable Features:   pprof-fp jemalloc mem-profiling portable sse test-engine-kv-rocksdb test-engine-raft-raft-engine cloud-aws cloud-gcp cloud-azure
Profile:           dist_release
./pd-server -V
 Release Version: v6.2.0-alpha
Edition: Community
Git Commit Hash: e1c6ad3d8b01ee875121c85006da8d1514a50fec
Git Branch: heads/refs/tags/v6.2.0-alpha
UTC Build Time:  2022-07-29 11:03:18

### What operating system and CPU are you using?
8core、16GB

### Steps to reproduce
https://tcms.pingcap.net/dashboard/executions/plan/1022596
6 tikv（5 replicas）
1、inject two tikv network isolation for 50min
2、recover network isolation

### What did you expect?
after recover network isolation，tikv status should be normal

### What did happened?
after the failure is restored, one tikv status still show down
origin_img_v2_35ba4dd1-ca21-4d08-8c8c-8ba0402809cg
image

### PR DESCRIPTION
**Pull request by bufferflies**
Signed-off-by: bufferflies <1045931706@qq.com>

### What is changed and how it works?

Issue Number: Close #13191

the RWR should be same in all os after https://github.com/rust-lang/rust/issues/93740， so the RWR deadlock   is not potential case in   linux. 

What's Changed:
Remove `call_option` avoid to read lock double.

```commit-message
```

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
- No code

TCMS Test：
https://tcms.pingcap.net/dashboard/executions/plan/1082168

image

Side effects

### Release note 

```release-note
remove call_option to avoid  deadlock(RWR).
```

### CODE REVIEW

**Review by lhy1024 — APPROVED** [review]

**Comment by bufferflies** [comment]
The pr: https://github.com/tikv/tikv/pull/13226 did not clear all RWR scene, sync_request will double read lock the inner:

When I delete pods tikv , the tikv can not exit , some stack trace is :

[the backgroud thread]
Thread 75 (Thread 0x7fe5719ca700 (LWP 32)):
#0  0x00007fe5839c71c9 in syscall () from /lib64/libc.so.6
#1  0x000055d3dbbbc9b8 in std::sys::unix::futex::futex_wait::h0c1abc2af9025d67 ()
#2  0x000055d3dbbab288 in std::thread::park::he5968c80ba8308e0 ()
#3  0x000055d3da56ce88 in futures_executor::local_pool::block_on::h28442ebcf6af778c ()
#4  0x000055d3da5c4886 in _$LT$pd_client..client..RpcClient$u20$as$u20$pd_client..PdClient$GT$::get_store::h07f480ec07eb50b0 ()
and log:

 [2022/08/09 04:05:39.919 +00:00] [INFO] [node.rs:520] ["stop raft store thred"] [store_id=6]

**Review by Connor1996 — APPROVED** [review]

File: components/pd_client/src/client.rs [hunk 1]
```diff
@@ -377,8 +363,8 @@ impl PdClient for RpcClient {
         let mut req = pdpb::IsBootstrappedRequest::default();
         req.set_header(self.header());
 
-        let resp = sync_request(&self.pd_client, LEADER_CHANGE_RETRY, |client| {
-            client.is_bootstrapped_opt(&req, Self::call_option(&self.pd_client))
+        let resp = sync_request(&self.pd_client, LEADER_CHANGE_RETRY, |client, call| {
```
Comment (nolouch) [inline components/pd_client/src/client.rs:366]: use a better name, like `|client, option|`.

File: components/pd_client/src/util.rs [hunk 2]
```diff
@@ -472,19 +472,26 @@ where
...
-    F: Fn(&PdClientStub) -> GrpcResult<R>,
+    F: Fn(&PdClientStub, CallOption) -> GrpcResult<R>,
 {
     loop {
         let ret = {
             // Drop the read lock immediately to prevent the deadlock between the caller
             // thread which may hold the read lock and wait for PD client thread
             // completing the request and the PD client thread which may block
             // on acquiring the write lock.
-            let client_stub = client.inner.rl().client_stub.clone();
```
Comment (youjiali1995) [inline components/pd_client/src/util.rs:486]: Did you mean the read lock here is stil held while getting `CallOption`?

Comment (youjiali1995) [reply components/pd_client/src/util.rs:486]: The read lock guard is dropped because of `clone`.

Comment (bufferflies) [reply components/pd_client/src/util.rs:486]: yes, I suspect this from the stack trace, the back ground task has wait for the futex ?

File: components/pd_client/src/client.rs (hunk 1 above)
Comment (bufferflies) [reply components/pd_client/src/client.rs]: fix

File: components/pd_client/src/util.rs [hunk 3]
```diff
@@ -472,19 +472,26 @@ where
...
-            func(&client_stub).map_err(Error::Grpc)
+            let inner = client.inner.rl();
```
Comment (youjiali1995) [inline components/pd_client/src/util.rs:493]: It's better to release the read lock before calling `func`.

Comment (bufferflies) [reply components/pd_client/src/util.rs:493]: fix

File: components/pd_client/src/util.rs [hunk 4]
```diff
@@ -472,19 +472,30 @@ where
...
         let ret = {
             // Drop the read lock immediately to prevent the deadlock between the caller
             // thread which may hold the read lock and wait for PD client thread
             // completing the request and the PD client thread which may block
             // on acquiring the write lock.
-            let client_stub = client.inner.rl().client_stub.clone();
-            func(&client_stub).map_err(Error::Grpc)
+            let (client_stub, option) = {
+                let inner = client.inner.rl();
+                return (inner.client_stub.clone(), call_option_inner(&inner));
```
Comment (youjiali1995) [inline components/pd_client/src/util.rs:495]:
```suggestion
                (inner.client_stub.clone(), call_option_inner(&inner))
```

Comment (bufferflies) [reply components/pd_client/src/util.rs]: fix

**Review by youjiali1995 — APPROVED** [review]

**Pull request merged**